Add HUC Parameter to Public Analyze, Model APIs

rajadain opened the pull request:
## Overview

We've used MMW-specific WKAoIs for a while for caching purposes, and as shorthand for established shapes. This serves the many kinds of in-built boundaries we support, including HUCs, School Districts, and Congressional Districts.

Now that we're opening up the API, for the convenience of our users, we now add the ability to specify a HUC in the request. HUCs are more widely known in the industry, and by letting users analyze and model them directly, without having to translate them into GeoJSON or WKAoI, will make the API a lot easier to use.

Connects #3474 

### Demo

#### Success

##### Analyze

```shell
xh --verbose POST :8000/api/analyze/soil/ Authorization:"Token b0c671e1424f37e58c81670ce8118e8eff0f0c14" huc==020402031008
```
```http
POST /api/analyze/soil/?huc=020402031008 HTTP/1.1
Accept: */*
Accept-Encoding: gzip, deflate, br
Authorization: Token b0c671e1424f37e58c81670ce8118e8eff0f0c14
Connection: keep-alive
Host: localhost:8000
User-Agent: xh/0.13.0

HTTP/1.1 200 OK
Allow: POST, OPTIONS
Connection: keep-alive
Content-Type: application/json
Date: Mon, 28 Feb 2022 22:24:29 GMT
Location: /api/jobs/9f947652-8515-4199-bf4b-8ae6e85a3547/
Server: nginx
Transfer-Encoding: chunked
Vary: Accept-Encoding
Vary: Accept, Cookie, Origin

{
    "job": "9f947652-8515-4199-bf4b-8ae6e85a3547",
    "status": "started"
}
```

```shell
xh --verbose :8000/api/jobs/9f947652-8515-4199-bf4b-8ae6e85a3547/ Authorization:"Token b0c671e1424f37e58c81670ce8118e8eff0f0c14"
```
```http
GET /api/jobs/9f947652-8515-4199-bf4b-8ae6e85a3547/ HTTP/1.1
Accept: */*
Accept-Encoding: gzip, deflate, br
Authorization: Token b0c671e1424f37e58c81670ce8118e8eff0f0c14
Connection: keep-alive
Host: localhost:8000
User-Agent: xh/0.13.0

HTTP/1.1 200 OK
Allow: GET, OPTIONS
Connection: keep-alive
Content-Type: application/json
Date: Mon, 28 Feb 2022 22:25:26 GMT
Server: nginx
Transfer-Encoding: chunked
Vary: Accept-Encoding
Vary: Accept, Cookie, Origin

{
    "job_uuid": "9f947652-8515-4199-bf4b-8ae6e85a3547",
    "status": "complete",
    "result": {
        "survey": {
            "name": "soil",
            "displayName": "Soil",
            "categories": [
                {
                    "area": 897.2539813519878,
                    "code": "a",
                    "coverage": 1.0505194818837915e-05,
                    "type": "A - High Infiltration"
                },
                {
                    "area": 3115265.823254102,
                    "code": "b",
                    "coverage": 0.036474036411005245,
                    "type": "B - Moderate Infiltration"
                },
                {
                    "area": 80847967.24370222,
                    "code": "c",
                    "coverage": 0.9465810843462092,
                    "type": "C - Slow Infiltration"
                },
                {
                    "area": 10767.047776223853,
                    "code": "d",
                    "coverage": 0.00012606233782605497,
                    "type": "D - Very Slow Infiltration"
                },
                {
                    "area": 0.0,
                    "code": "ad",
                    "coverage": 0.0,
                    "type": "A/D - High/Very Slow Infiltration"
                },
                {
                    "area": 151635.92284848593,
                    "code": "bd",
                    "coverage": 0.0017753779243836077,
                    "type": "B/D - Medium/Very Slow Infiltration"
                },
                {
                    "area": 1283970.4473146945,
                    "code": "cd",
                    "coverage": 0.015032933785757057,
                    "type": "C/D - Medium/Very Slow Infiltration"
                }
            ]
        }
    },
    "error": "",
    "started": "2022-02-28T22:24:29.689286Z",
    "finished": "2022-02-28T22:24:34.010561Z"
}
```

##### Model

```shell
xh --verbose :8000/api/modeling/gwlf-e/prepare/ Authorization:"Token b0c671e1424f37e58c81670ce8118e8eff0f0c14" huc=020402031008
```
```http
POST /api/modeling/gwlf-e/prepare/ HTTP/1.1
Accept: application/json, */*;q=0.5
Accept-Encoding: gzip, deflate, br
Authorization: Token b0c671e1424f37e58c81670ce8118e8eff0f0c14
Connection: keep-alive
Content-Length: 22
Content-Type: application/json
Host: localhost:8000
User-Agent: xh/0.13.0

{
    "huc": "020402031008"
}



HTTP/1.1 200 OK
Allow: POST, OPTIONS
Connection: keep-alive
Content-Type: application/json
Date: Mon, 28 Feb 2022 22:15:09 GMT
Location: /api/jobs/b660cbe4-ce47-49f0-915c-704a0a2104f1/
Server: nginx
Transfer-Encoding: chunked
Vary: Accept-Encoding
Vary: Accept, Cookie, Origin

{
    "job": "b660cbe4-ce47-49f0-915c-704a0a2104f1",
    "status": "started"
}
```

```shell
xh --verbose :8000/api/jobs/b660cbe4-ce47-49f0-915c-704a0a2104f1/ Authorization:"Token b0c671e1424f37e58c81670ce8118e8eff0f0c14"
```
```http
GET /api/jobs/b660cbe4-ce47-49f0-915c-704a0a2104f1/ HTTP/1.1
Accept: */*
Accept-Encoding: gzip, deflate, br
Authorization: Token b0c671e1424f37e58c81670ce8118e8eff0f0c14
Connection: keep-alive
Host: localhost:8000
User-Agent: xh/0.13.0

HTTP/1.1 200 OK
Allow: GET, OPTIONS
Connection: keep-alive
Content-Type: application/json
Date: Mon, 28 Feb 2022 22:16:17 GMT
Server: nginx
Transfer-Encoding: chunked
Vary: Accept-Encoding
Vary: Accept, Cookie, Origin

{
    "job_uuid": "b660cbe4-ce47-49f0-915c-704a0a2104f1",
    "status": "complete",
    "result": {
        "NRur": 10,
        "NUrb": 6,
        "TranVersionNo": "1.4.0",
        "SeepCoef": 0,
        "UnsatStor": 10,
        "SatStor": 0,
        "InitSnow": 0,
        "TileDrainRatio": 0,
        "TileDrainDensity": 0,
        "ETFlag": "<Hamon method>",
        ...
        },
    "error": "",
    "started": "2022-02-28T22:15:09.484303Z",
    "finished": "2022-02-28T22:15:14.107938Z"
}
```

#### Failure

##### Invalid HUC

```shell
xh --verbose :8000/api/modeling/gwlf-e/prepare/ Authorization:"Token b0c671e1424f37e58c81670ce8118e8eff0f0c14" huc=02040203100
```
```http
POST /api/modeling/gwlf-e/prepare/ HTTP/1.1
Accept: application/json, */*;q=0.5
Accept-Encoding: gzip, deflate, br
Authorization: Token b0c671e1424f37e58c81670ce8118e8eff0f0c14
Connection: keep-alive
Content-Length: 21
Content-Type: application/json
Host: localhost:8000
User-Agent: xh/0.13.0

{
    "huc": "02040203100"
}



HTTP/1.1 400 Bad Request
Allow: POST, OPTIONS
Connection: keep-alive
Content-Length: 65
Content-Type: application/json
Date: Mon, 28 Feb 2022 22:13:10 GMT
Server: nginx
Vary: Accept, Cookie, Origin

[
    "Specified HUC 02040203100 is 11 digits. Must be 8, 10, or 12."
]
```

##### HUC Not Found

```shell
xh --verbose :8000/api/modeling/gwlf-e/prepare/ Authorization:"Token b0c671e1424f37e58c81670ce8118e8eff0f0c14" huc=02040203100X
```
```http
POST /api/modeling/gwlf-e/prepare/ HTTP/1.1
Accept: application/json, */*;q=0.5
Accept-Encoding: gzip, deflate, br
Authorization: Token b0c671e1424f37e58c81670ce8118e8eff0f0c14
Connection: keep-alive
Content-Length: 22
Content-Type: application/json
Host: localhost:8000
User-Agent: xh/0.13.0

{
    "huc": "02040203100X"
}



HTTP/1.1 404 Not Found
Allow: POST, OPTIONS
Connection: keep-alive
Content-Type: application/json
Date: Mon, 28 Feb 2022 22:13:45 GMT
Server: nginx
Transfer-Encoding: chunked
Vary: Accept-Encoding
Vary: Accept, Cookie, Origin

{
    "detail": "No shape found for HUC 02040203100X"
}
```

### Notes

Initially we had assumed that the HUC lookups would be slow, and would require the addition of new indexes to make the queries performant. In practice, I found that was not the case. Since we use the same boundary tables in development as in staging and production, we should not see additional execution time in those environments either.

## Testing Instructions

* Check out this branch
* Test an Analyze endpoint with:
  - [x] GeoJSON input
  - [x] WKAoI input
  - [x] HUC input
* Test a Model endpoint with:
  - [x] GeoJSON input
  - [x] WKAoI input
  - [x] HUC input

rajadain commented on src/mmw/apps/geoprocessing_api/calcs.py:
```diff
@@ -366,3 +368,30 @@ def drexel_fast_zonal(geojson, key):
+                              'Must be 8, 10, or 12.')
+
+    if huc_len == 8:
+        huc_col = 'huc08'
```
This is because in some places (like the WKAoI `huc8__9`) we use `huc8`, and in others (like the table `boundary_huc08` and the column name `huc08`) we use `huc08`.

rajadain reviewed (commented):
Left a comment.

Also, the order of AoI precedence is:

- Use the GeoJSON if specified
- If not, but WKAoI is specified, use that to get GeoJSON, then use it
- If not, but HUC is specified, use that to get WKAoI, then use it

This will be further documented in #3488

emilyhu0106 reviewed (approved):
Having HUC params is very useful here! Nice and clean implementation.

A very small thing - I noticed that `wkaoi` and `huc` are query params in `analyze` while they are a part of the body object in `model`, which was a bit confusing when I was testing in postman, especially when the returned message is "Must supply either the area of interest (GeoJSON), or a WKAoI ID."

rajadain commented:
Updated that error message, and the general organization of the code, in [5486836](https://github.com/WikiWatershed/model-my-watershed/pull/3487/commits/54868360e74721bdf4a3f91d8ca7997af3a9f773). Updated the documentation in [d779026](https://github.com/WikiWatershed/model-my-watershed/pull/3487/commits/d779026bf926e7ff43f8e0f1707e869adb80fdc6).

I'll discuss the parameter placement with other engineers afterwards, and will make a change at that time. Going to merge this in once green.

Thanks for reviewing!
